pkg/querier/frontend2/frontend2.go

```diff
@@ -0,0 +1,290 @@
+	GRPCClientConfig  grpcclient.ConfigWithTLS `yaml:"grpc_client_config"`
+
+	// Used to find local IP address, that is sent to scheduler and querier-worker.
+	InfNames []string `yaml:"interface_names"`
+
+	// If set, address is not computed from interfaces.
+	Addr string `yaml:"address" doc:"hidden"`
+	Port int    `doc:"hidden"`
+}
+
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&cfg.SchedulerAddr, "frontend.scheduler-address", "", "DNS hostname used for finding schedulers.")
+	f.DurationVar(&cfg.DNSLookupPeriod, "frontend.scheduler-dns-lookup-period", 10*time.Second, "How often to query DNS.")
+	f.IntVar(&cfg.WorkerConcurrency, "frontend.scheduler-worker-concurrency", 5, "Number of goroutines pushing requests to ")
+
+	cfg.InfNames = []string{"eth0", "en0"}
+	f.Var((*flagext.StringSlice)(&cfg.InfNames), "frontend.interface", "Name of network interface to read address from.")
+	f.StringVar(&cfg.Addr, "frontend.address", "", "IP address to advertise to querier (via scheduler) (resolved via interfaces by default).")
+	f.IntVar(&cfg.Port, "frontend.port", 0, "Port to advertise to querier (via scheduler) (defaults to server.grpc-listen-port).")
+
+	cfg.GRPCClientConfig.RegisterFlagsWithPrefix("frontend.grpc-client-config", f)
+}
+
+// Frontend2 implements GrpcRoundTripper. It queues HTTP requests,
+// dispatches them to backends via gRPC, and handles retries for requests which failed.
+type Frontend2 struct {
+	services.Service
+
+	cfg Config
+	log log.Logger
+
+	lastQueryID atomic.Uint64
+
+	// frontend workers will read from this channel, and send request to scheduler.
+	requestsCh chan *frontendRequest
+
+	schedulerWorkers *frontendSchedulerWorkers
+	requests         *requestsInProgress
+}
+
+type frontendRequest struct {
+	queryID uint64
+	request *httpgrpc.HTTPRequest
+	userID  string
+
+	cancel context.CancelFunc
+
+	enqueue  chan enqueueResult
+	response chan *httpgrpc.HTTPResponse
+}
+
+type enqueueResult struct {
+	success bool // True if request was sent to scheduler successfully, and frontend should wait for response.
+	retry   bool // Whether request can be retried.
+
+	cancelCh chan<- uint64 // Channel that can be used for request cancellation. If nil, cancellation is not possible.
+}
+
+// New creates a new frontend.
+func NewFrontend2(cfg Config, log log.Logger, reg prometheus.Registerer) (*Frontend2, error) {
+	requestsCh := make(chan *frontendRequest)
+
+	schedulerWorkers, err := newFrontendSchedulerWorkers(cfg, fmt.Sprintf("%s:%d", cfg.Addr, cfg.Port), requestsCh, log)
+	if err != nil {
+		return nil, err
+	}
+
+	f := &Frontend2{
+		cfg:              cfg,
+		log:              log,
+		requestsCh:       requestsCh,
+		schedulerWorkers: schedulerWorkers,
+		requests:         newRequestsInProgress(),
+	}
+	// Randomize to avoid getting responses from queries sent before restart (which could lead to leak between tenants).
+	// This isn't perfect, but better than nothing.
+	f.lastQueryID.Store(rand.Uint64())
```

Comment:
It will just wrap back to 0 and continue from there. No component really cares about actual values. Frontend only uses it for pairing requests and responses received from queriers.